File: Cpp/MergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void merge(vector<int> &arr, int left, int mid, int right) {
    int i = left;
    int j = mid + 1;
    int k = 0;

    int *temp = new(nothrow) int[right - left + 1];

    if (!temp) {
        cout << "ERROR" << endl;
        return;
    }

    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j])temp[k++] = arr[i++];
        else temp[k++] = arr[j++];
    }
    while (i <= mid) temp[k++] = arr[i++];
    while (j <= right) temp[k++] = arr[j++];

    for(i=left, k=0; i<=right; i++,k++)arr[i]=temp[k];

    delete []temp;
}
void mergeSort(vector<int> &arr, int left, int right){
    if(left < right){
        int mid = left - (left - right) / 2;

        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        merge(arr,left,mid,right);
    }
}

File: Python/MergeSort.py
def mergeSort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        left = arr[:mid]
        right = arr[mid:]

        mergeSort(left)
        mergeSort(right)

        merge(left, right, arr)

def merge(left, right, arr):
    i = j = k = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            arr[k] += left[i]
            i += 1
        else:
            arr[k] += right[j]
            j += 1
        k += 1

    while i < len(left):
        arr[k] = left[i]
        i += 1
        k += 1

    while j < len(right):
        arr[k] = right[j]
        j += 1
        k += 1

list = [64, 34, 25, 12, 22, 11, 90]
print(mergeSort(list))